SIPS/sip-83.md

```diff
@@ -20,18 +20,42 @@ This SIP proposes to implement a mechanism to calculate and store a snapshot of
+The recomputation of the system debt pool will be managed by three distinct processes:
+
+1. A public function that will compute and save a snapshot of the total system debt at current prices.
+2. On each synth exchange, minting, or burning operation the debt snapshot will be updated with the debt deltas of the involved synths.
+3. Periodic recomputation of the overall debt by minters rather than by keepers.
+
+Together these will ensure that the on-chain total system debt does not diverge far from its true value without having
+to expensively recompute total system debt on every mint, burn, or transfer.
 
 ## Motivation
 
 <!--This is the problem statement. This is the *why* of the SIP. It should clearly explain *why* the current state of the protocol is inadequate.  It is critical that you explain *why* the change is needed, if the SIP proposes changing how something is calculated, you must address *why* the current calculation is innaccurate or wrong. This is not the place to describe how the SIP will address the issue!-->
 
-In order to enable minting, burning and claiming rewards the system needs to know the size of the debt pool and the user's collateral ratio, calculating this value is expensive and contributes to the cost of each mint and burn transaction. This is because each time the system must read the price and amount of every synth in the system, currently 40+.
+In order to enable minting, burning and claiming rewards the system needs to know the size of the debt pool and the
+user's collateral ratio, calculating this value is expensive and contributes to the cost of each mint and burn transaction.
+This is because each time the system must read the price and amount of every synth in the system, currently 40+.
 
-In the current gas environment minting, burning and claiming costs can reach \$50 USD+ per tx or higher, with the migration to external oracles this could rise by 50-100% (See https://sips.synthetix.io/sips/sip-84). This is due to the necessity of retrieving the latest prices and timestamp for each Synth from an external contract/s instead of reading all prices from our current `ExchangeRates` contract. About 70% of the gas costs are spent in calculating the total size of the debt pool during minting, burning and claiming rewards. Providing a snapshot of the total debt pool, calculated as:
+The current size of the debt pool is calculated as:
 
 \\[ Debt pool = \sum{Synths Total Supply * Synth Price}\\]
 
-allows subsequent minting and burning transactions to rely on the calculated debt pool size value. This would reduce the bottleneck of increasing gas costs when new synths are added and allow Synthetix protocol to onboard many more liquid synthetic assets such as sOIL and commodities in the future.
+Not only does this cost increase linearly with the number of synths in the system, it also necessitates retrieving
```

Comment:
By caching the total issued synths and not checking at the time of issuance, there is a possibility that the rates are actually stale but this would not be checked while using the previously cached total synths value. That being said, once the cache expired, then this window would be closed. It's something to keep in mind, and worth covering briefly in the SIP.

Reply:
This includes not just stale rates, but also invalid rates, as per [SIP-76](https://sips.synthetix.io/sips/sip-76)

Reply:
Good point. The implementation already addresses this by also caching the rates invalid state, which should certainly be reflected in the SIP.

Reply:
@justinjmoses  good point on invalid rates and could capture them as a single anyRatesInvalid() flag would be optimal ?